### PR TITLE
Allow clientoptions.token to be a string, tokendetails, or tokenrequest

### PR DESCRIPTION
**Pull request by SimonWoolf**
Fixes https://github.com/ably/docs/issues/75

Was surprised to find that `clientOptions.token` wasn't specified to accept anything other than a token string at the moment.

Question: now we're formally allowing `clientOptions.token` to be any of the three, should we deprecate `clientOptions.tokenDetails`? Cut down on redundant document clientoptions a bit. (If not, it's a bit inconsistent having `clientOptions.tokenDetails` but not `clientOptions.tokenRequest`)

### CODE REVIEW

**Comment by paddybyers** [comment]
> Question: now we're formally allowing clientOptions.token to be any of the three, should we deprecate clientOptions.tokenDetails? Cut down on redundant document clientoptions a bit. (If not, it's a bit inconsistent having clientOptions.tokenDetails but not clientOptions.tokenRequest)

We're not seriously suggesting this for typed languages are we?

**Comment by tcard** [comment]
> We're not seriously suggesting this for typed languages are we?

We do it already for the return value of `authCallback`, and I think it's cleaner than having different fields for things that have the same purpose.

In Objective-C we make TokenRequest, TokenDetails and String all implement a common interface that returns. Sadly in Java you can't implement interfaces for types you don't own, so I guess you have to take Object and downcast.

**Comment by paddybyers** [comment]
> I guess you have to take Object and downcast.

That's so bad I think a better API, in the java case, is to have differently named and typed fields.

**Comment by mattheworiordan** [comment]
> That's so bad I think a better API, in the java case, is to have differently named and typed fields.

I think for ~~crappy~~ limited languages like Java, we should indeed simply deviate where necessary.  But I don't think the entire API should suffer as a result of primarily Java.

**Comment by mattheworiordan** [comment]
I think this should be based on `source-0-10`, see https://github.com/ably/wiki/issues/166#issuecomment-262910738

**Comment by SimonWoolf** [comment]
rebased on source-0-10

File: content/client-lib-development-guide/features.textile [hunk 1]
```diff
@@ -1025,7 +1025,7 @@ h4. ClientOptions
 ** @(TO3i)@ @recover@ string - A connection recovery string, specified with the intention of inheriting the state of an earlier connection
 ** @(TO3j)@ Auth option attributes:
 *** @(TO3j1)@ @key@ string - Full Ably key string as obtained from dashboard
-*** @(TO3j2)@ @token@ string - An authentication token string issued for this application
+*** @(TO3j2)@ @token@ string | @TokenDetails@ | @TokenRequest@ - An authentication token issued for this application, either in the form of a token string, a @TokenDetails@ object, or a @TokenRequest@ object
```
Comment (SimonWoolf) [inline content/client-lib-development-guide/features.textile]: Given Paddy's objection, I guess maybe something like this?
`@(TO3j2)@ @token@ string | @TokenDetails@ | @TokenRequest@ - An authentication token issued for this application, either in the form of a token string, a @TokenDetails@ object, or a @TokenRequest@ object. For languages where it's impractical or unidiomatic to take multiple types, this should take a token string only, and TO3j3 and TO3j12 apply`

`(TO3j3) tokenDetails TokenDetails – for languages where @token@ (TO3j2) takes a string only, this is a @TokenDetails@ object`

`(TO3j12) tokenRequest TokenRequest – for languages where @token@ (TO3j2) takes a string only, this is a @TokenRequest@ object`

Comment (mattheworiordan) [inline content/client-lib-development-guide/features.textile]: LGTM 👍

**Pull request closed without merging**